### ISSUE TITLE
Establish Symfony coding standards for private readonly properties and readonly classes

### ISSUE THREAD

**Issue by ghost** [symfony/symfony#57602]
### Description

AFAIR, when Symfony bumped the minimum PHP version to 8.1 and people started using `private readonly` properties in pull requests the response from the core team was negative. I even remember a comment from @nicolas-grekas who said that private readonly properties don't make sense.
Now I can see that we have 257 occurences https://github.com/search?q=repo%3Asymfony%2Fsymfony+%22private+readonly%22&type=code.

I fully agree that private readonly properties do not make sense and only make visual noise. I think people don't understand why this keyword was introduced to the language (I bet 99% don't read RFCs, only New Features section :smile:) and started using it everywhere when they expect a property to be set once, doesn't matter it's a private api, so this keyword became more like a validation/protection use case.
What's more, 99.9% of private readonly properties are dependencies. Have you ever override a property that holds a dependency? 

The same goes for readonly classes. Readonly on the class indicates a class is, surprise, **readonly**. But it makes sense only for structures like DTOs, VO etc. For classes that **do a job**, i.e. so called services, the readonly keyword doesn't make sense. You don't read anything from this class because they are not designed for reading (by "reading" I mean reading properties, not methods that return some data).

I think having a standard would be nice **at least** for consistency, otherwise some code will have private readonly properties and some not.
Please add a comment if you vote :-1:.

P.S. I bet the readonly keyword would have never been introduced to the PHP if we has asymmetric visibility already.

Reference:
NIcolas' feedback on private readonly properties:
https://github.com/symfony/symfony/pull/46674#discussion_r897618119
https://github.com/symfony/symfony/pull/53279#discussion_r1439248703

**Comment by derrabus** [comment]
Sorry, but your issue description is filled with speculation about how "people" don't understand certain language concepts. I don't want to participate in those speculations, nor do I find them relevant for this project.

I'd like to close this issue.

**Comment by ghost** [comment]
@derrabus So you understand this language concept differently? Please elaborate.

Also I would have thought twice before creating this issue if Nicolas hadn't added this comment I remember until today. Taking into account his authority I think there must be something to this topic.

btw. I'm not surprised you voted "no" because I saw many private readonly properties in doctrine/dbal for example.

**Comment by derrabus** [comment]
> Please elaborate.

No.

**Comment by ghost** [comment]
> No.

So please delete your previous comment if you don't want to take a part in the discussion.

P.S. I'm sorry to say that but from all the core team members that I respect and like, you are the one that I place in "not so kind" box. If I had to point one person from the core team that needs to work on personal skills I would point you. And I'm saying this based on watching your communication with community since you became a core team member.

**Comment by derrabus** [comment]
> > No.
> 
> So please delete your previous comment

I will do no such thing. I think I made clear why I disapproved of your opening comment and that disapproval will stay for the record.

> P.S. I'm sorry to say that but from all the core team members that I respect and like, you are the one that I place in "not so kind" box. If I had to point one person from the core team that needs to work on personal skills I would point you. And I'm saying this based on watching your communication with community since you became a core team member.

I don't know why you feel like you need to take this to the personal level here. Also, I don't think that your personal feelings towards me are relevant for the topic that you allegedly wanted to discuss.

That being said, I do support the initiative to…

> Establish Symfony coding standards for private readonly properties and readonly classes

I believe we need some kind of guidelines on when and how we want to use `readonly` in this codebase. So, if someone were to start a discussion on this topic in a respectful and objective way, I'm sure I'd participate in that discussion.

**Comment by ro0NL** [comment]
> private readonly properties don't make sense

i like we cant reset/mutate services/deps.. whats wrong with that?

**Comment by ghost** [comment]
> i like we cant reset/mutate services/deps.. whats wrong with that?

What is the thing is a trade off. How many times in you life have you overwritten a private property that you haven't supposed to? I've been in PHP world for many years and I've never heard about such a mistake. Also haven't seen a single person asking for feature to protect against overriding private properties. Thus a question is whether we want having all the codebase with readonly keyword everywhere or it's an uncecessary noise that doesn't bring much value if any.  

For me marking private property as readonly is like setting a password on a password, where `private` is the first password (nobody but me can set this property) and `readonly` is the second password (even though it's private I will add another level of security to protect from myself so I don't set it twice by accident).

**Comment by wouterj** [comment]
Take in mind that the [Symfony Code of Conduct](https://symfony.com/coc) is in effect to GitHub issues. This is no place for (unconstructive) personal attacks to anybody, we're having a purely technical discussion.

---

I personally don't feel the need for a strict standard here. Things like this can be decided on a case-by-case basis.

I also can't remember Nicolas comment nor do I see a reference to it. What would the downsides of going with readonly properties?

**Comment by ghost** [comment]
@wouterj I hope this note about Code of Conduct was directed to derrabus who first violated the CoC.

> Things like this can be decided on a case-by-case basis.

I don't think the readonly keyword next to the private property is case-by-case specific. This is a standard or not. I also don't think this is something that developers keep an eye on when doing code reviews so it will end up being a mess eventually.

> I also can't remember Nicolas comment

I have finally found this comment and even second comment. Issue description updated. 

> What would the downsides of going with readonly properties?

I think I've explained this in issue description and in this comment https://github.com/symfony/symfony/issues/57602#issuecomment-2198349160

**Comment by wouterj** [comment]
The comments tell us that they don't add enough value to make it worth fighting merge conflicts for 2 years. So changing existing properties is probably something we don't want.

However, I don't see any arguments for not adding this to new properties/code.

---

Personally, I only see a personal attack from you in this issue. Alexander merely stated that he is against this issue and doesn't want to elaborate on that (which is in his full rights).
But if you feel like the CoC is violated, these matters can be handled through the CARE team and not through an attack in the comments.

**Comment by wouterj** [comment]
To further support this, the first hit in the search result is a property introduced by Nicolas: https://github.com/symfony/symfony/pull/48642/files#diff-b24b2d134d203cc7851340a7155af5639befdbe019fcf3bcd8bcf270b1122254R26

So I think our general stance is yes for new properties, no for existing properties (which we may change once all support branches require PHP 8.1+).

**Comment by ghost** [comment]
> To further support this, the first hit in the search result is a property introduced by Nicolas:

Yes, I have seen that. If you look at the code review in that PR you will find out why Nicolas added readonly and how he commented on that (the emoji is essential) https://github.com/symfony/symfony/pull/48642#discussion_r1052467363.
Let's wait until tomorrow as Nicolas seems not to work on the weekends and I'm really curious what his statment will be on this discussion.

**Comment by wouterj** [comment]
Going further, at least 6 different core team members introduced readonly private properties: [#48362](https://github.com/symfony/symfony/pull/48362#discussion_r1033745593), [#54432](https://github.com/symfony/symfony/pull/54432/files#diff-ec21ff779ff549cec297da26af87d7d0c66ca302474e5fdad4bc7618d1c290a2R26), [#47112](https://github.com/symfony/symfony/pull/47112/commits/a18127b78910426be230704dc623fc3ad8f07614#diff-7975001ac86cbde62c42b008982254403f6c4e66a68d96bf1044c076e8dea093R28), [#51169](https://github.com/symfony/symfony/pull/51169/files#diff-c1bf609d9399cd70b79d086c40e6a257ccd74aeb1319a24be30a68300c2ce050R29), [#54589](https://github.com/symfony/symfony/pull/54589/files#diff-9b104ce3b114cd49a68f5f2c1dd0304e0a1ea7501092498e23852e159c2ed1cfR33), [#50112](https://github.com/symfony/symfony/commit/e71a3a1ce515ae2d8af5342de2c1677875a6cd87#diff-a00430a1bd69df9ce8939da70168ef066a25fe212435e23c3c028547f60594abR21)

But sure, let's leave this open for a bit longer to see if other core team members want to confirm the opinion.

**Comment by nicolas-grekas** [comment]
I'm not comfortable with this issue at all. It makes things way too personal. My supposed authority is not an argument on its own. And derrabus just stated his opinion - then you made it personal.

On the topic itself: I don't like readonly props because they are flawed by design (or at least too many think they provide immutability, while that's just false). But they do have merits, so in the end: I don't care. It would be just me, I wouldn't use them much. And also: it would be just me, Symfony wouldn't be as great.

In short: I have other topics more important to care about and given the polemic nature of such topics (not to say the general tone of the discussion), I wouldn't like to enforce any rule on the specific issue of private readonly.

**Comment by wouterj** [comment]
I agree. Let's close here then.

**Comment by ghost** [comment]
1. At first derrabus accused me that I speculate how people understand certain language concepts when what I said was that people don't understand **why this keyword was introduced to the language** (the wrong or correct undestanding the private readonly case is another topic where I only gave my personal opinion). Even besides the wording that could have been wrongly understood, his comment was not appropriate because it doesn't apply to the reported issue itself but tries to judge what derrabus thinks I think.
2. Then I asked him to elaborate on his understanding of readonly keyword. He's unswer was single `No` which is a CoD violation IMO. He didn't want to talk on the issue but he wanted to insinuate me the speculations. This was disrespectful behaviour IMO.
3. My personal opinion about derrabus was a consequence of the two points above.

**Comment by ghost** [comment]
> or at least too many think they provide immutability

@derrabus why don't you accuse Nicolas with speculation about how "people" don't understand certain language concepts? All developers are equal but some are more equal than others? Would you ever write this first comment if Nicolas was an author of this issue? This is what I had in mind when I wrote about you being not so kind. I feel like your attitude to people in conversations depends on how you respect/know this person. I don't experience such feeling when talking with another core team members (or reading conversations I'm not a part of) . Please take what I said into consideration because it doesn't encourage people to get involved into open source.

Best.

**Comment by wouterj** [comment]
Again, please follow through the CARE team if you think action is required. I'll lock down this issue to prevent any more deviation from the *technical* discussions that need to happen on GitHub.
